### PR TITLE
feat: Toggle visibility of 360 image markers

### PR DESCRIPTION
**Pull request by astrid-kg**
#### Type of change
Feat  
Bug 

#### Jira ticket :blue_book:
https://cognitedata.atlassian.net/browse/REV-843

## Description :pencil:
Exposed `setIconsVisibility` and made a debug UI for it in Examples. However, this feature/function was not working. We were changing visibility during icon culling, so any changed in visibility will be overwritten in `IconCollection` by either `setIconClustersByLOD` or `computeProximityPoints` on next update.

This was fixed by distinguishing between culled and visible. An icon is only visible if it is **not culled AND set as visible**. If an icon is invisible, it will not be added during culling.

**Note**: We currently have a bug where the icon that is entered should be hidden during transition, but due to the culling scheme this is no longer working (maybe it never did). This is also fixed by this PR. We also have a super rare bug where the icon you entered is "covering the screen" and consumes input. This will probably also be fixed by being able to set the entered icon as invisible.

## How has this been tested? :mag:
On Windows in Chrome. I have tested the new "hide all" UI, but also tested culling in general by entering/leaving 360 images.

## Test instructions :information_source:
Load a set of 360 images:
- Hibernia_RS2
- helideck-site-2-jpeg
- helideck-site-2

1. Click icons to enter and use esc to leave. Make sure that the icon you left are still visible, and that culling looks correct in both states.
2. Use the "Hide all 360 Images" in the "360 Images" UI. Try this both outside and inside 360 images. Make sure that icons are only visible when they should be.

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [x] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [ ] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: viewer/packages/api/src/api-helpers/Image360ApiHelper.ts [hunk 1]
```diff
@@ -294,7 +294,7 @@ export class Image360ApiHelper {
   }
 
   public exit360Image(): void {
-    this._image360Facade.allIconsVisibility = true;
+    this._interactionState.currentImage360Entered?.icon.setVisibility(true);
```
Comment (haakonflatval-cognite) [inline viewer/packages/api/src/api-helpers/Image360ApiHelper.ts:297]: If the 360 image collection is hidden when this happens, the icon will be visible while the rest are hidden.

**Review by haakonflatval-cognite — CHANGES_REQUESTED** [review]
Looks good - just a couple of small issues I'd like to see sorted out before merge

File: examples/src/utils/Image360UI.ts [hunk 2]
```diff
@@ -78,6 +79,18 @@ export class Image360UI {
...
+      .add(iconCulling, 'hideAll')
+      .name('Hide all 360 images')
+      .onChange(() => {
+        if (sets.length > 0) {
+          sets.forEach(p => p.setIconsVisibility(!iconCulling.hideAll));
+          viewer.requestRedraw();
+        } else {
+          iconCulling.hideAll = !iconCulling.hideAll;
+        }
+      });
```
[the remark this answers is not in the document]
Comment (astrid-kg) [reply examples/src/utils/Image360UI.ts:90]: Done! Added `collection.set360CollectionVisibility(!iconCulling.hideAll)` to `add360ImageSet()`! 🍔

File: examples/src/utils/Image360UI.ts [hunk 3]
```diff
@@ -78,6 +79,18 @@ export class Image360UI {
...
 
+    gui
+      .add(iconCulling, 'hideAll')
+      .name('Hide all 360 images')
+      .onChange(() => {
+        if (sets.length > 0) {
+          sets.forEach(p => p.setIconsVisibility(!iconCulling.hideAll));
+          viewer.requestRedraw();
+        } else {
+          iconCulling.hideAll = !iconCulling.hideAll;
```
[the remark this answers is not in the document]
Comment (astrid-kg) [reply examples/src/utils/Image360UI.ts]: This was intentional, but now that I `set360CollectionVisibility` in `add360ImageSet` I have remove it. 🌊

File: viewer/packages/api/src/api-helpers/Image360ApiHelper.ts (hunk 1 above)
Comment (astrid-kg) [reply viewer/packages/api/src/api-helpers/Image360ApiHelper.ts]: Fixed: See https://github.com/cognitedata/reveal/pull/3074/commits/951c27062f3171e6b7aa67dbc863d857165bde6f

File: viewer/packages/360-images/src/Image360Facade.ts [hunk 4]
```diff
@@ -50,23 +50,28 @@ export class Image360Facade<T> {
...
   public preload(entity: Image360Entity, lockDownload?: boolean): Promise<void> {
     return this._image360Cache.cachedPreload(entity, lockDownload);
   }
 
+  public getCollectionContainingEntity(entity: Image360Entity): DefaultImage360Collection {
+    const imageCollection = this._image360Collections.filter(collection =>
+      collection.image360Entities.includes(entity)
+    );
+    if (imageCollection.length !== 1) {
+      throw new Error(`360Image entity contained in invalid number of collections: {${imageCollection.length}}`);
```
Comment (haakonflatval-cognite) [inline viewer/packages/360-images/src/Image360Facade.ts:70]:
```suggestion
      throw new Error(`360Image entity contained an invalid number of collections: {${imageCollection.length}}`);
```

File: viewer/packages/360-images/src/collection/DefaultImage360Collection.ts [hunk 5]
```diff
@@ -104,7 +119,7 @@ export class DefaultImage360Collection implements Image360Collection {
   }
 
   public setIconsVisibility(visible: boolean): void {
```
Comment (haakonflatval-cognite) [inline viewer/packages/360-images/src/collection/DefaultImage360Collection.ts:121]:
```suggestion
  private setIconsVisibility(visible: boolean): void {
```
This shouldn't be used outside this class anymore, right?

**Review by haakonflatval-cognite — APPROVED** [review]
Minor nitpick - otherwise, LGTM :+1: :tada:

File: viewer/packages/api/src/api-helpers/Image360ApiHelper.ts (hunk 1 above)
Comment (haakonflatval-cognite) [reply viewer/packages/api/src/api-helpers/Image360ApiHelper.ts]: Nice!

File: viewer/packages/360-images/src/Image360Facade.ts (hunk 4 above)
Comment (astrid-kg) [reply viewer/packages/360-images/src/Image360Facade.ts]: Changed to: "Failed to get Collection for Image360Entity. The entity is present in ${imageCollection.length} collections."

File: viewer/packages/360-images/src/collection/DefaultImage360Collection.ts (hunk 5 above)
Comment (astrid-kg) [reply viewer/packages/360-images/src/collection/DefaultImage360Collection.ts]: Recombined the functions. We now only have exposed setIconsVisibility.

**Pull request merged**